### ISSUE TITLE
Fixing SAC Policy

### ISSUE THREAD

**Issue by gggoes** [JuliaReinforcementLearning/ReinforcementLearning.jl#970]
I believe SAC is still an open issue, as the current implementation is unable to learn a viable policy. Even the JuliaRL_SAC_Pendulum [example](https://juliareinforcementlearning.org/docs/experiments/experiments/Policy%20Gradient/JuliaRL_SAC_Pendulum/#JuliaRL\\_SAC\\_Pendulum) does not show signs of improvement as the number of episodes grows. I've been reading the original [paper](https://arxiv.org/abs/1812.05905) for the past days and I believe I found the reason for this behaviour and the way to fix it.

The problem, I think, is not the SAC code itself, but the GaussianNetwork implementation (available in networks.jl in main branch, neural_network_approximator.jl in my version). Here is the [current](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/main/src/ReinforcementLearningCore/src/utils/networks.jl) implementation

```
function (model::GaussianNetwork)(rng::AbstractRNG, s; is_sampling::Bool=false, is_return_log_prob::Bool=false)
    x = model.pre(s)
    μ, raw_σ = model.μ(x), model.σ(x)
    σ = clamp.(raw_σ, model.min_σ, model.max_σ)
    if is_sampling
        z = ignore_derivatives() do
            noise = randn(rng, Float32, size(μ))
            μ .+ σ .* noise
        end
        if is_return_log_prob
            logp_π = sum(normlogpdf(μ, σ, z) .- (2.0f0 .* (log(2.0f0) .- z .- softplus.(-2.0f0 .* z))), dims=1)
            return z, logp_π
        else
            return z
        end
    else
        return μ, σ
    end
end
```
After carefully reading SAC details, I made a couple of changes to this code. Here is my version.
```
function (model::GaussianNetwork)(rng::AbstractRNG, s; is_sampling::Bool=false, is_return_log_prob::Bool=false)
    x = model.pre(s)
    μ, raw_logσ = model.μ(x), model.logσ(x)
    logσ = clamp.(raw_logσ, log(model.min_σ), log(model.max_σ))
    if is_sampling
        σ = exp.(logσ)
        noise = randn(rng, Float32, size(μ))
        z = μ .+ σ .* noise
	action = tanh.(z)
        if is_return_log_prob
	    ϵ = 1.0f-10        
	    logp_π = sum(normlogpdf(μ, σ, z) .- log.(1 .- action .^ 2 .+ ϵ), dims=1)
            return action, logp_π
        else
            return action
        end
    else
        return μ, logσ
    end
end
```
Here are the changes I made

1.  I used a squashing function, so the action output is action = tanh.(z), where z = μ .+ σ .* noise. See Appendix C. That means the environments needs to map the [-1,1] value to the real action, but that doesn't seem to be a major constraint.
2.  I changed logp_π, see equation 26. Added a small constant for stability. Honestly I have no idea where the other equation came from.
3.  Removed the ignore block. I am not sure about why, as I am not very family with Zygote, but results showed me this part is important. Possibly because it helps to implement something similar to equation 10

The σ = exp.(logσ) part is obsolete now, but that is not really important to the general idea. Normalizer field also seems to be unimportant.

I ran my own environment after these changes and results were very good, close to the results I got with DDPG. α, the temperature, was always positive (I noticed it was negative sometimes with the old implementation) and its value seems to decrease with the number of episodes. That shows the policy is becoming more deterministic and that the agent is learning. 

For a real implementation, μ last layer should not be relu, as it only outputs positive numbers and action will be always positive (unless noise is very negative). I simply used identity. 

PPO also employs Gaussian Networks. I am not familiar with its code so I am not sure how these changes would impact it. Nevertheless, I am confident that is the correct implementation for SAC. Feedback is welcome.

**Comment by HenriDeh** [comment]
Hello, thanks for your work on that. I'm not saying this in a

**Comment by HenriDeh** [comment]
Hello, thanks for willing to contribute to RL.jl. It is I who redesigned GaussianNetwork the way it is currently. I'm happy to see people interested in improving it and in fixing SAC. Your suggestion is very reminiscent of the way GaussianNetwork used to be before I reworked it. Here are some remarks on your post:

1. I don't think GaussianNetwork is broken (I'm not claiming it is fully correct, though) because it works just fine with the MPO algorithm. 
2. I removed the formerly existing tanh "activation" function for two reasons; 1) it is mathematically incorrect as explained in #745 and 2) it is in fact a major constraint for some applications of RL other than robotics and games. For this latter reason, I moved it to a struct field (normalizer) so that it is only optional. 
3. I don't know either where the old function comes from. It seemed correct when I compared its results with that which is implemented in Distributions.jl. You can also check out diagmvnormlogpdf which could simply replace this. 
4. The ignore block signals that Zygote is not supposed to back propagate through the sampling, but only through the logpdf function. This fact is important, as explained in #598. Your remark is interesting though because it seems indeed that in Equation (10), the sampling is back propagated through. Which means that we may need to differentiate both cases depending on the needs of specific algorithms. PPO for instance I'm pretty sure shouldn't backprop through the sampling either. 
5. I also expressly removed the exp in #839 because it made the network training unstable. Perhaps you don't experience this issue because you work with the tanh normalizer. 

There's definitely some improvement that can be done here. After all, you did see SAC improve with your changes. But as you can see it's more involved than you may think.

**Comment by gggoes** [comment]
Thank you for replying. Since posting this issue, I had more time to test this new implementation. Here are my results with PendulumEnv, using the same code as JuliaRL_SAC_Pendulum (just more steps). The agent is definitely learning.

![sac_implementation](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/assets/141772411/56c06f16-e0a7-4031-b776-6eaf1619681e)

Final temperature α is 0.0259, so the policy becomes very deterministic. Policy entropy is -3.22, which is very close to the target = - dim(A) = -3.  (I believe p.entropy_term should be = - entropy = - mean(log_π), but it doesn't make a lot of difference since it is only logging)
```

        reward = mean(q)
        entropy = mean(log_π)
        ignore() do
            p.reward_term = reward
            p.entropy_term = entropy
        end
```

Just to add my own remarks on top of your remarks:

1. Yes, perhaps I worded it badly. I don't think the GaussianNetwork implementation itself is wrong, I just don't think it is compatible with SAC. It clearly works very well with PPO, but given the differences between PPO and SAC algorithms its performance varies greatly.

2. Honestly I didn't understand 2.1. If X,Y are random variables and Y = f(X), X ~ Normal(μ, σ) and f = tanh, then Equation (28) is the logical result to [obtain](https://en.wikipedia.org/wiki/Random_variable#Functions_of_random_variables) the pdf of y. As for 2.2, I believe you mean that some environments shouldn't be constrained to the [-1,1] range? Maybe in that case, tanh shouldn't be used and the log probability should be returned simply as logp_π = sum(normlogpdf(μ, σ, z)), where z is now the action. A bounded_action::Bool field could be included to differentiate these cases. Either way, logp_π should be consistent with the squashing function (or lack thereof)	

3. It took me a while but I finally understood it. 1 - tanh(z)^2 = 1 / cosh(z)^2. So log(1 - tanh(z)^2) = -2 log(cosh(z)). Using cosh(z) = (exp(z) + exp(-z))/2, then finally we get 2 * (log(2) - z - log(1+exp(-2*z))). It is really the same logp_π ! Readability is not great, but I guess who implemented wrote that way so as to avoid numerical instabilities when z is very large. That also means tanh needs to be used to be consistent with this probability.

4. I am very confident Zygote should read the sampling part of the code. I made the other 2 changes first and performance didn't change much. Only after removing ignore_derivative did it improve. Equation 10 after all defines the policy update.

5. I had no problem so far using logσ but again I don't have enough experience to comment on that. 

So I'd argue the only necessary changes are action = tanh.(z) and remove the ignore block. I guess the initial implementation was almost correct, except that the ignore was added so there would be no conflict with PPO (thus degrading SAC)

**Comment by HenriDeh** [comment]
According to the [spinning up documentation](https://spinningup.openai.com/en/latest/algorithms/sac.html#spinup.sac_pytorch) of open AI, the reparametrization should indeed be differentiable:
![image](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/assets/47037088/9a0f31b6-4471-47e0-8c6e-33a79f89688d)
And apparently they use another closed-form formula, given in the appendix, to compute the mvnormlogpdf when a tanh is used:
![image](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/assets/47037088/8c52aa4d-35a1-4a0f-bfbf-67a455517830)

So it seems a special policy network is needed to fix SAC, along with a dedicated mvnormlogpdf function. It can use the log-std output to reproduce the paper. Regarding tanh, I'm not sure if it is mandatory for SAC to work, but for simplicity we can content ourselves with the paper version and upgrade to an optional variant if someone needs it one day. Would you be willing to contribute it by making a PR ? I'll help if necessary and review it.

**Comment by gggoes** [comment]
Yes, I am willing to contribute. Creating a new network for SAC should be very straightforward, by adapting the contents of GaussianNetwork.
